core/src/main/java/org/infinispan/factories/InternalCacheFactory.java

```diff
@@ -167,4 +209,296 @@ private void bootstrap(String cacheName, AdvancedCache<?, ?> cache, Configuratio
+            newCache.internalWire(this);
+            return newCache;
+         }
+         return this;
+      }
+
+      @Override
+      public AdvancedCache<K, V> withFlags(Collection<Flag> flags) {
+         long newFlags = EnumUtil.bitSetOf(flags);
+         long updatedFlags = EnumUtil.mergeBitSets(bitFlags, newFlags);
+         if (bitFlags != updatedFlags) {
+            PartitionHandlingCache<K, V> newCache = new PartitionHandlingCache<>(super.withFlags(flags), updatedFlags);
+            newCache.internalWire(this);
+            return newCache;
+         }
+         return this;
+      }
+
+      @Override
+      public AdvancedCache<K, V> noFlags() {
+         if (bitFlags != 0) {
+            PartitionHandlingCache<K, V> newCache = new PartitionHandlingCache<>(super.noFlags(), 0L);
+            newCache.internalWire(this);
+            return newCache;
+         }
+         return this;
+      }
+   }
+
+   private static class StatsCache<K, V> extends AbstractGetAdvancedCache<K, V, StatsCache<K, V>> {
+
+      @Inject private TimeService timeService;
+      private CacheMgmtInterceptor interceptor;
+
+      public StatsCache(AbstractGetAdvancedCache<K, V, ?> cache) {
+         this((AdvancedCache<K, V>) cache);
+      }
+
+      private StatsCache(AdvancedCache<K, V> cache) {
+         super(cache, new AdvancedCacheWrapper<K, V>() {
+            @Override
+            public AdvancedCache<K, V> wrap(AdvancedCache<K, V> cache) {
+               throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public AdvancedCache<K, V> wrap(AdvancedCache<K, V> self, AdvancedCache<K, V> newDelegate) {
+               StatsCache<K, V> newCache = new StatsCache<>(newDelegate);
+               newCache.internalWire((StatsCache<K, V>) self);
+               newCache.interceptorStart();
+               return newCache;
+            }
+         });
+      }
+
+      @Override
+      protected void internalWire(StatsCache<K, V> cache) {
+         this.timeService = cache.timeService;
+         super.internalWire(cache);
+      }
+
+      private void interceptorStart() {
+         // This has to be done after the cache is wired - otherwise we can get a circular dependency
+         interceptor = cache.getAsyncInterceptorChain().findInterceptorWithClass(CacheMgmtInterceptor.class);
+      }
+
+      @Override
+      public V get(Object key) {
+         V value;
+         if (interceptor == null) {
+            interceptorStart();
+         }
+         if (interceptor.getStatisticsEnabled()) {
+            long beginTime = timeService.time();
+            value = cache.get(key);
+            interceptor.addDataRead(value != null, timeService.timeDuration(beginTime, TimeUnit.NANOSECONDS));
+         } else {
+            value = cache.get(key);
+         }
+         return value;
+      }
+   }
+
+   private static class GetReplCache<K, V> extends AbstractGetAdvancedCache<K, V, GetReplCache<K, V>> {
```

Comment:
It should yes, I mention on the original https://issues.jboss.org/browse/ISPN-9212 that we could even do these for DIST/scattered in a different way. I would say we can add that in later if need though.

Reply:
You can probably do it for dist/scattered as well if you check the segment first - otherwise you might hit a stale entry that you no longer own but haven't removed yet after the last topology update. But we'd probably have to do it within `CacheImpl`, where you can compute the segment, instead of a `CacheImpl` wrapper.

I called out local because the conditions and implementation should be exactly the same.

Reply:
Yeah I agree it should be completely the same for LOCAL/INVAL.